### PR TITLE
Peer in pushpin

### PR DESCRIPTION
**Pull request by mjtognetti**
Adds a Storage Peer content type which can be used to store data in a remote storage peer.

### CODE REVIEW

File: src/renderer/components/content-types/storage-peer/StoragePeer.css [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  justify-content: center;
+  align-items: center;
+  width: 100%;
+  box-sizing: border-box;
+  padding: 10px;
+  background-color: white;
+  border: 1px solid var(--colorPaleGrey);
+}
+
+.StoragePeer--boardLabel {
```
Comment (mjtognetti) [inline src/renderer/components/content-types/storage-peer/StoragePeer.css]: This isn't used anywhere

File: src/renderer/components/content-types/storage-peer/StoragePeerWorkspace.css [hunk 2]
```diff
@@ -0,0 +1,60 @@
+.StoragePeerEditor {
+  color: var(--colorBlueBlack);
+}
+
+.StoragePeerEditor-frame {
```
Comment (mjtognetti) [inline src/renderer/components/content-types/storage-peer/StoragePeerWorkspace.css]: In SUIT convention, `<Component>-<subComponent>` indicates a subcomponent, but in the markup `.StoragePeerEditor-frame` wraps `.StoragePeerEditor`. It doesn't look like `.StoragePeerEditor` does anything other than set `color`, so you could collapse these two classes into just `.StoragePeerFrame` and remove the extra div from the markup.

File: src/renderer/components/content-types/storage-peer/StoragePeerWorkspace.tsx [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+  const { hypermergeUrl } = props
+  const [doc, changeDoc] = useDocument<StoragePeerDoc>(hypermergeUrl)
+  const selfId = useSelfId()
+  const workspaceUrlsContext = useContext(WorkspaceUrlsContext)
+  const [selfDoc, changeSelfDoc] = useDocument<ContactDoc>(selfId)
+
+  if (!workspaceUrlsContext) {
+    return null
+  }
+  const currentWorkspace = workspaceUrlsContext.workspaceUrls[0]
```
Comment (mjtognetti) [inline src/renderer/components/content-types/storage-peer/StoragePeerWorkspace.tsx]: We should implement a `useCurrentWorkspace()` hook which hides this logic so components don't need to know about the ordering of workspaces. Noting for a separate patch.

File: src/renderer/components/content-types/storage-peer/StoragePeerWorkspace.tsx [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  const registerWithStoragePeer = useCallback(() => {
+    changeDoc((doc) => {
+      doc.storedUrls[selfId] = workspaceUrl
+    })
+
+    changeSelfDoc((selfDoc) => {
+      if (!selfDoc.devices) {
+        selfDoc.devices = []
+      }
+      if (doc && !selfDoc.devices.includes(hypermergeUrl)) {
```
Comment (mjtognetti) [inline src/renderer/components/content-types/storage-peer/StoragePeerWorkspace.tsx]: why check `doc` here?

File: src/renderer/components/content-types/storage-peer/StoragePeerWorkspace.tsx [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    })
+  }, [selfDoc])
+
+  if (!doc) {
+    return null
+  }
+
+  const { storedUrls } = doc
+
+  if (!storedUrls && (doc as any).archivedUrls) {
```
Comment (mjtognetti) [inline src/renderer/components/content-types/storage-peer/StoragePeerWorkspace.tsx]: We should be able to get rid of this `archivedUrls` check.

File: src/renderer/components/content-types/storage-peer/StoragePeerWorkspace.tsx [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    changeDoc((doc) => {
+      doc.storedUrls = (doc as any).archivedUrls
+    })
+  }
+
+  const archiveEntries = Object.entries(storedUrls)
```
Comment (mjtognetti) [inline src/renderer/components/content-types/storage-peer/StoragePeerWorkspace.tsx]: name update: `storedEntries`

File: src/renderer/components/content-types/storage-peer/index.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
+import ContentTypes from '../../../ContentTypes'
+import { HypermergeUrl } from '../../../ShareLink'
+
+import StoragePeerWorkspace from './StoragePeerWorkspace'
+import StoragePeer from './StoragePeer'
+import { DeviceDoc } from '../workspace/Device'
+
+export interface StoragePeerDoc extends DeviceDoc {
+  storedUrls: { [contact: string]: HypermergeUrl } // contact is actually a hypermergeURL too
```
Comment (mjtognetti) [inline src/renderer/components/content-types/storage-peer/index.ts]: A convention for noting this kind of information we've been using in hypermerge is:
`storedUrls: { [contact: string /*HypermergeUrl*/]: HypermergeUrl }`

File: src/renderer/components/content-types/storage-peer/StoragePeerWorkspace.tsx (hunk 5 above)
Comment (pvh) [reply src/renderer/components/content-types/storage-peer/StoragePeerWorkspace.tsx]: yup, migration's done ;)

File: src/renderer/components/content-types/storage-peer/StoragePeerWorkspace.css (hunk 2 above)
Comment (pvh) [reply src/renderer/components/content-types/storage-peer/StoragePeerWorkspace.css]: yea, i blindly took the contact editor CSS so i'll fix the other one too

File: src/renderer/components/content-types/storage-peer/StoragePeer.css (hunk 1 above)
Comment (pvh) [reply src/renderer/components/content-types/storage-peer/StoragePeer.css]: cut

**Pull request merged**